include/nbl/builtin/hlsl/bxdf/reflection/lambertian.hlsl

```diff
@@ -0,0 +1,176 @@
+{
+    using this_t = LambertianParams<LS, SI, Scalar>;
+
+    static this_t create(NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(SI) interaction, BxDFClampMode _clamp)
+    {
+        this_t retval;
+        retval._sample = _sample;
+        retval.interaction = interaction;
+        retval._clamp = _clamp;
+        return retval;
+    }
+
+    // iso
+    Scalar getNdotV() NBL_CONST_MEMBER_FUNC { return hlsl::mix(math::conditionalAbsOrMax<Scalar>(_clamp == BxDFClampMode::BCM_ABS, interaction.getNdotV(), 0.0), interaction.getNdotV(), _clamp == BxDFClampMode::BCM_NONE); }
+    Scalar getNdotVUnclamped() NBL_CONST_MEMBER_FUNC { return interaction.getNdotV(); }
+    Scalar getNdotV2() NBL_CONST_MEMBER_FUNC { return interaction.getNdotV2(); }
+    Scalar getNdotL() NBL_CONST_MEMBER_FUNC { return hlsl::mix(math::conditionalAbsOrMax<Scalar>(_clamp == BxDFClampMode::BCM_ABS, _sample.getNdotL(), 0.0), _sample.getNdotL(), _clamp == BxDFClampMode::BCM_NONE); }
+    Scalar getNdotLUnclamped() NBL_CONST_MEMBER_FUNC { return _sample.getNdotL(); }
+    Scalar getNdotL2() NBL_CONST_MEMBER_FUNC { return _sample.getNdotL2(); }
+    Scalar getVdotL() NBL_CONST_MEMBER_FUNC { return _sample.getVdotL(); }
+
+    // aniso
+    Scalar getTdotL2() NBL_CONST_MEMBER_FUNC { return _sample.getTdotL() * _sample.getTdotL(); }
+    Scalar getBdotL2() NBL_CONST_MEMBER_FUNC { return _sample.getBdotL() * _sample.getBdotL(); }
+    Scalar getTdotV2() NBL_CONST_MEMBER_FUNC { return interaction.getTdotV() * interaction.getTdotV(); }
+    Scalar getBdotV2() NBL_CONST_MEMBER_FUNC { return interaction.getBdotV() * interaction.getBdotV(); }
+
+    LS _sample;
+    SI interaction;
+    BxDFClampMode _clamp;
+};
+
+template<class LS, class Iso, class Aniso, class Spectrum NBL_PRIMARY_REQUIRES(LightSample<LS> && surface_interactions::Isotropic<Iso> && surface_interactions::Anisotropic<Aniso>)
+struct SLambertianBxDF
+{
+    using this_t = SLambertianBxDF<LS, Iso, Aniso, Spectrum>;
+    using scalar_type = typename LS::scalar_type;
+    using ray_dir_info_type = typename LS::ray_dir_info_type;
+    using isotropic_interaction_type = Iso;
+    using anisotropic_interaction_type = Aniso;
+    using sample_type = LS;
+    using spectral_type = Spectrum;
+    using quotient_pdf_type = sampling::quotient_and_pdf<spectral_type, scalar_type>;
+
+    using params_isotropic_t = LambertianParams<LS, Iso, scalar_type>;
+    using params_anisotropic_t = LambertianParams<LS, Aniso, scalar_type>;
+
+
+    static this_t create()
+    {
+        this_t retval;
+        // nothing here, just keeping in convention with others
+        return retval;
+    }
+
+    static this_t create(NBL_CONST_REF_ARG(SBxDFCreationParams<scalar_type, spectral_type>) params)
+    {
+        return create();
+    }
+
+    void init(NBL_CONST_REF_ARG(SBxDFCreationParams<scalar_type, spectral_type>) params)
+    {
+        // do nothing
+    }
```

Comment:
if you go the route of "BxDFs have special parameter structs" then use composition and keep a `params` member (which has its own `create` factory) then you have no `create` in the BxDF itself and you expect someone to set the parameters to comething valid

Reply:
This is turning a bit of a mess because you need two typedefs `query_params` and `creation_params` which should be two separate things:
- query params are the interaction, sample, cache, etc. for a single ray
- creation params are constant for the whole material, such as roughness

You can't have a single `SBxDFCreationParams` to cover all BxDFs